File: src/dotnet/Src/ButterflyNet.Parser/ScopeTree.cs
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;

namespace ButterflyNet.Parser {

	public class ScopeTree {

		private readonly ICollection<ScopeTreeNode> nodes = new Collection<ScopeTreeNode>();

		public void AddNode(ScopeTreeNode node) {
			nodes.Add(node);
		}

		public IEnumerable<ScopeTreeNode> Nodes { get { return nodes; } }
		public int Count { get { return nodes.Count + nodes.Sum(node => node.Count); } }

		public override string ToString() {
			return string.Format("ScopeTree(Count={0})", Count);
		}
	}

	public static class ScopeTreeExtensions {
		public static ScopeTreeNode GetMostRecentNode(this ScopeTree tree, int depth) {
			return depth > 0 && tree.Nodes.Any() ? GetMostRecentNode(tree.Nodes, depth) : null;
		}

		private static ScopeTreeNode GetMostRecentNode(IEnumerable<ScopeTreeNode> nodes, int depth, int currentDepth = 0) {
			var node = nodes.Last();
			return currentDepth < depth && node.Children.Any() ? GetMostRecentNode(node.Children, depth, ++currentDepth) : node;
		}
	}

}

File: src/js/src/ScopeTree.js
function getCount(nodes) {
	var count = nodes.length;
	for (var i = 0; i < nodes.length; i++) {
		count += nodes[i].count();
	}
	
	return count;
}

function ScopeTree() {
	this.nodes = [];
	
	this.addNode = function(node) {
		this.nodes.push(node);
	};
	
	this.isEmpty = function() { return this.nodes.length === 0; };
	this.count = function() { return getCount(this.nodes); };
}

function ScopeTreeNode(scope) {
	this.scope = scope;
	this.parent = null;
	this.children = [];
	
	this.addChild = function(node) {
		node.parent = this;
		this.children.push(node);
	};
	
	this.count = function() {
		return getCount(this.children);
	};
	
	this.depth = function() {
		return 1 + (this.parent ? this.parent.depth() : 0);
	};
}

ScopeTree.prototype = {
	getMostRecentNode: function() {
		function getMostRecentNode(nodes, depth, currentDepth) {
			var node = nodes[nodes.length - 1];
			return currentDepth < depth && node.children.length > 0 ? getMostRecentNode(node.children, depth, ++currentDepth) : node;
		}
		
		return function(depth) {
			return depth > 0 && !this.isEmpty() ? getMostRecentNode(this.nodes, depth, 0) : null;
		};
	}()
};